Make H5MD serialize with pickle

edisj opened the pull request:
Fixes #2890 

Changes made in this Pull Request:
 - Added `H5PYPicklable` class to `H5MDReader`


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

pep8speaks commented:
Hello @edisj! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/coordinates/H5MD.py`](https://github.com/MDAnalysis/mdanalysis/blob/5bcd6c12bdcbdab57f4cc350c5df3cda95fc450d/package/MDAnalysis/coordinates/H5MD.py):

> [Line 560:80](https://github.com/MDAnalysis/mdanalysis/blob/5bcd6c12bdcbdab57f4cc350c5df3cda95fc450d/package/MDAnalysis/coordinates/H5MD.py#L560): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

edisj commented:
Hi @orbeckst, hopefully this fixes the pickling issues. I followed the code @yuxuanzhuang gave in #2890 which does allow the tests to pass. I modified it a bit to work with `comm` arguments as well and added documentation. 

It seems to work with some testing:
![image](https://user-images.githubusercontent.com/59893899/89764876-1af85800-daaa-11ea-9b30-d9bd13ef8e81.png)
as in the `h5py.File` object seems to have all the data in it that it should, but I'm not entirely confident with how pickling works. Please leave any comments/criticisms for changes

yuxuanzhuang commented:
The Travis error can be solved by adding `mock`, like in `chemfiles`:
```python
try:
    import h5py
except ImportError:
    HAS_H5PY = False

    # Allow building documentation even if h5py is not installed
    import imp

    class MockH5pyFile:
        pass
    h5py = imp.new_module("h5py")
    h5py.File = MockH5pyFile

else:
    HAS_H5PY = True
```

yuxuanzhuang commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +728,93 @@ def has_forces(self):
+            driver = self._driver
+            comm = self._comm
+        except AttributeError:  # is this error necessary?
+            driver = None
```
Will that be a problem if `self._driver` exists but `self._comm` not?

edisj replied:
You're right I think it would have been a problem. `self._driver` should have been `self.driver` as well, since it's pulling the driver from the file object, right? The pickled file wasn't returning the right driver with `self._driver`, but does with `self.driver`

yuxuanzhuang commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +728,93 @@ def has_forces(self):
+        self.__dict__ = state
+        self._particle_group = self._file['particles'][
+                                list(self._file['particles'])[0]]
+
```
```suggestion
        self[self.ts.frame]

```
This is not needed in this specific case, but just to fit to be broader API.

orbeckst commented:
@yuxuanzhuang thank you for reviewing.

@edisj please ping me when all tests are passing and when you have addressed all of @yuxuanzhuang 's comments. I can then do a final review + approval.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +740,89 @@ def has_forces(self):
+        except AttributeError:
+            driver = None
+
+        return {'name': self.filename,
```
I am confused: If the `comm` is not included in `__getstate__` then how can it be used in `__setstate__`?

orbeckst replied:
I see that you had it included in b29e675765c71ee0e015b3690e6b38a902a86188 but then removed again.

Did you test that a `MPI.COMM_WORLD` could be pickled on its own, i.e.,
```python
from mpi4py import MPI
import pickle

comm = MPI.COMM_WORLD
comm_p = pickle.loads(pickle.dumps(comm))

assert comm == comm_p
```
or something similar?

edisj replied:
That does not seem to work
![image](https://user-images.githubusercontent.com/59893899/89835031-c50ec900-db18-11ea-9f8a-d199c2789f89.png)

yuxuanzhuang replied:
I don’t have a working mpi4py in my desktop. But it seems that it cannot be pickled natively.
This might help meanwhile https://bitbucket.org/mpi4py/mpi4py/issues/104/pickling-of-mpi-comm

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +740,89 @@ def has_forces(self):
+            self.__init__(name=state['name'],
+                          mode=state['mode'],
+                          driver=state['driver'],
+                          comm=state['comm'])
```
Where will `state['comm']` come from?

edisj replied:
I thought it was taking the `state` dict as argument from https://github.com/MDAnalysis/mdanalysis/blob/c7087b8cb1473b98ca943057d1697b1a3f91a46d/package/MDAnalysis/coordinates/H5MD.py#L744-L747

orbeckst replied:
Maybe I don't understand how pickling works but I assumed that the goal is to make each object picklable by itself, i.e., ,`H5PYPicklableFile` **by itself** can be pickled. That's the first thing to check (and add a test for!!). Once this works, `H5MDReader` can just treat it as one item in its own class dict, as it would a numpy array or a basic Python object. It would be bad program design if `H5PYPicklableFile` needed the help of `H5MDReader` because then you could never use it independently from `H5MDReader`.

edisj replied:
I see, that makes sense, thanks for the clarification. I'm doing some testing with `comm` on the parallel build I have on spudda

orbeckst commented:
@edisj please see #2897 and switch imp to using importlib.

edisj commented:
I'm making a lot of commits at the moment because I'm pulling them to my repository on workstation for testing with parallel h5py

orbeckst commented:
Don't worry, we can squash it all when merging.

edisj commented:
@orbeckst The latest commit successfully pickles `h5py.File` objects with `driver='mpio` and `comm=MPI.COMM_WORLD`, although I'm still unsure if the correct communicator is being unpickled. I started a conversation in the h5py group and will continue to poke around online in the meantime.

![image](https://user-images.githubusercontent.com/59893899/89863473-032edb80-db5f-11ea-8927-e327c5a32ef2.png)

yuxuanzhuang commented:
I am trying to build a parallel h5py system. You could add a few tests (specifically for H5PYPicklable) inside ``testsuite/MDAnalysisTests/utils/test_pickleio.py`` to increase codecov. (add skipif marks for them)

yuxuanzhuang commented:
I am having trouble to install all that stuffs to make h5py work. But I assume if it is hard to find the "right" `comm`, you might  overwrite the `__init__` function to save the input `comm` with something like
```python
def __init__(self, ...comm):
   self._comm = comm
   super().__init__(...)
```

orbeckst commented:
The test failures are pretty annoying for all other PRs. Can we make H5PYPicklable work for _serial_ right now? It would be ok to raise a `TypeError` (or whatever else is normally raised) if a communicator is present. Add a test that checks that `comm != None` raises the error.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +757,112 @@ def has_forces(self):
+    def __getstate__(self):
+        driver = self.driver
+        if driver == 'mpio':
+            comm = self.id.get_access_plist().get_fapl_mpio()[0]
```
If `MPI.COMM_WORLD` cannot be natively pickled then I don't think that this will work.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +757,112 @@ def has_forces(self):
+
+    def __getstate__(self):
+        driver = self.driver
+        if driver == 'mpio':
```
In order to move the PR along, make `__getstate__` fail (with `TypeError`) when you have driver mpio and a comm object.

Add a test that specifically checks for the exception being raised. Then open a new issue to make pickling of parallel h5py files work.

edisj replied:
The test fails with `ModuleNotFoundError` because h5py tries to import mpi4py when `driver='mpio'`. Is there any way around this?

orbeckst replied:
Which test are we talking about?

edisj replied:
something like this:
```
@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
def test_H5MD_parallel_pickle():
    from unittest.mock import MagicMock
    MPI_COMM_WORLD = MagicMock()
    h5md_io = H5PYPicklable(H5MD_xvf, 'r',
                            driver="mpio",
                            comm=MPI_COMM_WORLD)
    h5md_io_pickled = pickle.loads(pickle.dumps(h5md_io))
```

orbeckst replied:
Try mocking the mpi4py module, too.  The idea is to provide a thing (the mock) that just pretends to be there but doesn't really do anything.

orbeckst replied:
@richardjgowers you have been dubbed in other PR "The King of Mocks" – any suggestions?

yuxuanzhuang replied:
I assume it cannot be fixed as it's checked inside h5py code. (https://github.com/h5py/h5py/blob/8cc83410995996330c85867d0e56cbe922b709f8/h5py/_hl/files.py#L43)

Maybe just excluding these lines from coverage for now?

orbeckst replied:
There might be a way to trick `h5py` into believing that it was built with mpi support (one could patch the call to `h5.get_config()` https://github.com/h5py/h5py/blob/8cc83410995996330c85867d0e56cbe922b709f8/h5py/_hl/files.py#L24 or just the `mpi` variable) but that's not worth it for right now so I concur with @yuxuanzhuang to just skip the test. Add a comment summarizing the current issues and move forward.

edisj replied:
@orbeckst  just pushed commit with the lines excluded from coverage

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -727,3 +757,112 @@ def has_forces(self):
+        return ()
+
+
+def comm_unpickle(name):
+    return getattr(MPI, name)
+
+
+def comm_pickle(obj):
+    if obj == MPI.COMM_NULL:
+        return comm_unpickle, ('COMM_NULL',)
+    if obj == MPI.COMM_SELF:
+        return comm_unpickle, ('COMM_SELF',)
+    if obj == MPI.COMM_WORLD:
+        return comm_unpickle, ('COMM_WORLD',)
+    if isinstance(obj, MPI.Comm):
+        return comm_unpickle, ('COMM_WORLD',)
+    raise TypeError("cannot pickle object")
+
+
+if HAS_MPI:
+    copyreg.pickle(MPI.Intracomm, comm_pickle, comm_unpickle)
+    copyreg.pickle(MPI.Comm, comm_pickle, comm_unpickle)
```
Remove for the serial-pickle PR

orbeckst commented:
See https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/370755668#L2255-L2257
```
    def test_read_charges(self, top):
>       assert_almost_equal(top.charges.values, 0.0)
E       NameError: name 'assert_almost_equal' is not defined
testsuite/MDAnalysisTests/topology/test_hoomdxml.py:82: NameError
```

orbeckst commented:
@yuxuanzhuang could you please review this PR within the next day? Sorry for asking for a quick turnaround but @edisj 's project is coming to an end and we would like to have it at a good stopping point, just in case he has other more urgent things to do. 

You should be able to leave a approve/request changes review.  I will wait with merging until I hear from you.

orbeckst commented:
Thank you @edisj and @yuxuanzhuang !
